---
bibliography: C:\\Users\\John\\Documents\\library.bib
csl: C:\\Users\\John\\Documents\\ecology.csl
output:
  bookdown::word_document2:
    reference_docx: C:\\Users\\John\\Documents\\rmd_style_template_single_spaced.docx
    toc: no
  bookdown::html_document2:
    code_folding: hide
    fig_caption: yes
    theme: united
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
---

```{r, include=FALSE, cache = FALSE}

knitr::opts_chunk$set(cache = FALSE)

#initialize manuscript formatting script
source('C://Users//John//Documents//initialize_manuscript.R')

```

# Disturbance and recovery of mouse gut microbiota in response to antibiotics.

\newline

John Guittar^1,\*^, Aspen Reese^2^, Lawrence David^3^, Timothy Brennhofer^1,4^, Ashley Shade^5,6^, Elena Litchman^1,6,7^

\newline

^1^ Kellogg Biological Station, Michigan State University, 3700 E Gull Lake Dr, Hickory Corners, MI 49060, USA

^2^ Harvard University, Cambridge, MA 02138, USA

^3^ Department of Molecular Genetics and Microbiology, Duke University, Box 3382, Durham, 11 NC 27708, USA

^4^ Grinnell College, 1115 8th Ave, Grinnell, IA 50112, USA

^5^ Department of Microbiology and Molecular Genetics, Department of Plant, Soil and Microbial Sciences, and The Plant Resilience Institute, Michigan State University, East Lansing MI 48840, USA

^6^ Program in Ecology, Evolutionary Biology and Behavior, Michigan State University East Lansing MI 48840, USA

^7^ Department of Integrative Biology, Michigan State University, East Lansing, MI 48824, USA

^\*^ Correspondence author: E-mail: guittarj@msu.edu

\newline

#Abstract

\newline

#Introduction

* We want to improve our understanding of antibiotic disturbance and recovery in the mammalian gut. 
* Most work in this area has focused on shifts in abundance at coarse taxonomic scales. We want to look at disturbance and recovery at the OTU scale, in order to more explicitly examine ecological dynamics.
* One way to better understand ecological mechanisms is to focus on traits. For example, in plant systems, traits related to dispersal ability and rapid growth are often associated with early colonists to distubed areas. Indeed, these traits often form a syndrome characteristic of pioneer species.
* Colonization of the gut after disturbance my be particularly difficult, because it is an abiotic community with strong impediments to dispersal. Spores have been hypothesized to be a life history trait that enhances the probability of dispersal among hosts.

\newline

##*Questions:*

1. Which traits (other than antibiotic resistance) predict OTU tolerance/susceptibility to antibiotics?

2. Which traits predict recovery after antibiotic disturbances?

3. Which traits predict successful immigration from the regional pool during the recovery process?

`r indent` - Are these traits related to dispersal ability and rapid growth, as expected?

\newline

To answer these questions, we will use data from @Reese2018.

\newline

Another way to frame many of these questions is: can OTU traits help to predict the boxes into which OTUs fall in the following diagram:

![](images/winners_losers.png)

`r ref('figure', 'figOTUFates')`

(ref:figOTUFates) Figure `r refs$figOTUFates`

**(ref:figOTUFates):** The potential fates and fortunes of OTUs in mice treated with antibiotics. Positive or neutral outcomes are shaded gray, whereas negative outcomes are shaded red.

`r pagebreak`

![](images/setup_Reese2018.png)

`r ref('figure', 'figReese2018Setup')`

(ref:figReese2018Setup) Figure `r refs$figReese2018Setup`

**(ref:figReese2018Setup):** A figure drawn from @Reese2018 showing the experimental setup and data collection schedule.

`r pagebreak`

\newline

# Methods

\newline

# Results

```{r, include=FALSE}

# load packages and custom functions
wd <- 'C:\\Users\\John\\Documents\\msu\\Mouse_Microbiome\\'
wd_dat <- 'C:\\Users\\John\\Google Drive\\Mouse_Microbiome_Shared_Files\\Data\\'
setwd(wd)

source(paste0(wd, 'custom_functions.R'))
loadpax(pkg = c('tidyverse','knitr','data.table','vegan', 'ape'))

#set directory for printed images/figures
image_dir <- paste0(wd, 'images\\')

# load data
x_raw <- fread(paste0(wd_dat, '2018_06_12_run2_otus.csv'))
tax <- fread(paste0(wd_dat, 'otus_table_joined_taxa.txt'))
traits <- readRDS(paste0(wd_dat, 'traits_castor_subtree_averaging.RDS')) %>%
  mutate(otu = as.integer(gsub('otu', '', otu)))
 
#Organize tax data into traditional classifications
tax <- tax %>%
  rename(otu = `#OTU ID`, tax = taxonomy) %>%
  mutate(tax = gsub("\\s|.__|\\[|\\]|Other", "", tax)) %>%
  separate(tax, sep = ';', c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species"), fill = 'right')
tax <- as.data.frame(apply(tax, 2, function(x) ifelse(x == '', 'unclassified', x)), stringsAsFactors = FALSE)
tax$otu <- as.numeric(tax$otu)

if(FALSE) {
  saveRDS(tax, file = 'C:\\Users\\John\\Google Drive\\Mouse_Microbiome_Shared_Files\\Data\\Reese2018_OTUs_wide.RDS')
}

# Note that, oddly, x[130,4719] is 'NA'. I'm assuming it is zero...?
x <- as.data.frame(x_raw)
is.na(x[130,4719])
x[130,4719] <- 0

#remove samples with fewer than 5000 reads.
x <- x[rowSums(x[c(11:ncol(x))]) >= 5000, ]

#remove cage 310 because mouse dies
#x <- x[x$`Cage#` != 310, ]

#ensure all OTUs occur at least once after removing samples
x <- bind_cols(x[, c(1:10)], x[, c(11:ncol(x))[colSums(x[,c(11:ncol(x))]) > 0]])

#save unrarified version of data?
x_unrarified <- x

#rarefy to 5000
tmp <- x[,c(11:ncol(x))]
tmp <- tmp[rowSums(tmp) >= 5000,]
tmp <- rrarefy(tmp, 5000)
x <- bind_cols(x[, c(1:10)], as.data.frame(tmp))

#create wide table for distance calculations
x <- x %>%
  mutate(Mouse = floor(SampleID)) %>%
  select(-Day, -Run, -Reads, -`Cage#`, -SampleID)

#fix data entry problems
x$Cohoused[x$Mouse == 3 & x$Hour == 168] <- 'N'
x$Cohoused[x$Mouse == 6 & x$Hour == 168] <- 'N'
x$Cohoused[x$Mouse == 7 & x$Hour == 168] <- 'Y'
x$Pair[x$Mouse == 7 & x$Hour == 168] <- 4
x$Abx[x$Mouse == 7 & x$Hour == 168] <- 'Y'
x$Dose[x$Mouse == 7 & x$Hour == 168] <- 'Recovery'
x$Cohoused[x$Mouse == 9 & x$Hour == 168] <- 'Y'
x$Pair[x$Mouse == 9 & x$Hour == 168] <- 2
x$Abx[x$Mouse == 9 & x$Hour == 168] <- 'N'
x$Dose[x$Mouse == 9 & x$Hour == 168] <- 'No'
x$Cohoused[x$Mouse == 12 & x$Hour == 168] <- 'N'
x$Cohoused[x$Mouse == 16 & x$Hour == 168] <- 'N'

#save wide table for later
x_wide <- x

if (FALSE) {
  saveRDS(x_wide, file = 'C:\\Users\\John\\Google Drive\\Mouse_Microbiome_Shared_Files\\Data\\Reese2018_OTUs_wide.RDS')
}


#cleanup, make narrow table
# I could relativize to a percent here, but why?
x <- x %>%
  gather(otu, abun, -Mouse, -Abx, -Hour, -Cohoused, -Dose, -Pair) %>%
  group_by(Mouse, Abx, Hour, Cohoused, Dose, Pair, otu) %>%
  summarise(abun = sum(abun)) %>%
  mutate(otu = as.numeric(otu)) %>%
  group_by(otu) %>%
  filter(sum(abun) > 0) %>%
  ungroup()
  
#filter so trait and tax tables have the same taxa.
#create wide version
traits <- filter(traits, otu %in% x$otu)
traits_wide <- spread(traits, trait, val, fill = NA)
tax <- filter(tax, otu %in% x$otu)

#set up phyloseq object for UniFrac distances.
#Want to do it but can't until I make a tree...

#Set plotting theme
th <- theme_bw() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.background = element_blank())

#set trait names for plotting
#renaming
trait_names <- c(
  "B_vitamins" = "B vitamins",
  "Copies_16S" = "16S gene copies",
  "Forms_biofilms" = "Forms biofilms",
  "GC_content" = "GC content",
  "Gene_number" = "Genes",
  "Gram_positive" = "Gram-positive",
  "IgA" = "IgA binding affinity",
  "Obligate_anaerobe" = "Obligate anaerobe",
  "Temp_optimum" = "Temperature optimum"
)


```

```{r, include = FALSE}

#####
#takes too long
#ord <- metaMDS(x_wide, k = 3, trymax = 1000, noshare = 0.2)
#####

myp <- select(x_wide, -Mouse, -Abx, -Dose, -Hour, -Cohoused, -Pair)
myp <- pcoa(vegdist(myp))

j <- data.frame(
  x_wide[, c('Mouse','Abx','Hour','Pair')], 
  PCoA1 = myp$vectors[,1],
  PCoA2 = myp$vectors[,2]) %>%
  group_by(Mouse) %>%
  mutate(
  label = ifelse(Abx == 'Y' & (Hour == min(Hour) | Hour == 120 | Hour == max(Hour)), Hour, NA),
  Abx = factor(ifelse(Abx == 'Y','Abx', 'No abx'), levels = c('No abx', 'Abx')))

p1 <- j %>% 
  filter(!is.na(Pair)) %>%
  mutate(Pair = paste('Cage', Pair)) %>%
  ggplot(aes(x = PCoA1, y = PCoA2, color = Abx, fill = Abx)) + 
  geom_point(aes(alpha = Hour), shape = 21, size = 4) + 
  geom_point(shape = 21, size = 4, fill = NA) + 
  geom_path() + 
  geom_text(aes(label = label), color = 'black', na.rm=TRUE) +
  scale_color_manual(values = c('black','red')) +
  scale_fill_manual(values = c('black','red')) +
  facet_grid(Pair~Abx) + 
  th

plot_it(p1, name = 'figPCoAPaired', dir = image_dir, width = 7, height = 8)

p1

```

(ref:figPCoAPaired) Figure `r refs$figPCoAPaired`

`r pagebreak` ![](images\\figPCoAPaired.png)

**(ref:figPCoAPaired):** PCoA of mouse microbiomes over time, during and after 120 hours of antibiotic treatment. Control mice (left) and treated mice (right) were cohoused after treatment (hour 120) for the duration of the experiment.

```{r, include = FALSE}

p2 <- j %>% 
  filter(is.na(Pair)) %>%
  group_by(Abx) %>%
  mutate(Pair2 = as.numeric(as.factor(Mouse))) %>%
  ggplot(aes(x = PCoA1, y = PCoA2, color = Abx, fill = Abx)) + 
  geom_point(aes(alpha = Hour), shape = 21, size = 4) + 
  geom_point(shape = 21, size = 4, fill = NA) + 
  geom_path() + 
  geom_text(aes(label = label), color = 'black', na.rm=TRUE) +
  scale_color_manual(values = c('black','red')) +
  scale_fill_manual(values = c('black','red')) +
  facet_grid(Pair2~Abx) + th

plot_it(p2, name = 'figPCoASolitary', dir = image_dir, width = 7, height = 8)

p2

```

(ref:figPCoASolitary) Figure `r refs$figPCoASolitary`

`r pagebreak` ![](images\\figPCoASolitary.png)

**(ref:figPCoASolitary):** PCoA of mouse microbiomes over time, during and after 120 hours of antibiotic treatment. Control mice (left) and treated mice (right) were not cohoused at any point during the experiment. 


```{r, include = FALSE}

j <- tax %>% 
  gather(tax_level, name, -otu) %>% 
  full_join(x, by = 'otu') %>% 
  filter(Hour <= 120) %>%
  filter(tax_level != 'Domain') %>% 
  filter(abun > 0)

j <- j %>% 
  mutate(tax_level = factor(tax_level, levels = c('Phylum','Class','Order','Family','Genus','Species'))) %>%
  group_by(tax_level) %>% 
  mutate(total_rich = n_distinct(name)) %>% 
  group_by(Mouse, tax_level) %>% 
  mutate(rich = n_distinct(name), val = rich / total_rich,
         Abx = ifelse(Abx == 'Y', 'Abx', 'No abx'))

p1 <- j %>% 
  ggplot(aes(x = tax_level, y = val, group = Mouse, color = Abx)) + 
  #geom_line() + 
  stat_summary(fun.data = "mean_sdl", fun.args = list(mult=1), aes(group = Abx), position = position_dodge(width = 0.2)) +
  scale_color_manual(values = c('black','red')) +
  expand_limits(y = c(0,1)) +
  labs(x = 'Taxonomic level', y = 'Percent of all Phyla present across all mice') +
  th

plot_it(p1, name = 'figTaxaSharedAcrossMice', dir = image_dir, width = 5, height = 4)

p1

```

(ref:figTaxaSharedAcrossMice) Figure `r refs$figTaxaSharedAcrossMice`

`r pagebreak` ![](images\\figTaxaSharedAcrossMice.png)

**(ref:figTaxaSharedAcrossMice):** Percent of phyla observed across all mice present within single mice, assessed at increasingly specific taxonomic levels. 

```{r, include = FALSE}

j <- x %>%
  group_by(Mouse, Hour) %>%
  mutate(rich = length(unique(otu[abun > 0]))) %>%
  group_by(Mouse, Hour, Abx, Cohoused, Dose, Pair, rich) %>%
  do(shannon = diversity(spread(select(., otu, abun), otu, abun))) %>%
  mutate(shannon_val = shannon[[1]])

#must fix screwy data!!! (when I paste mouse, abx, cohoused, there are some unexpected combinations)
j <- j %>%
  group_by(Mouse, Abx, Cohoused) %>%
  filter(length(shannon_val) > 1) %>%
  ungroup() %>%
  mutate(
    evenness = shannon_val / log(rich),
    Abx = ifelse(Abx == 'Y', 'Abx', 'No abx'),
    Mouse = ifelse(Cohoused == 'Y', paste(Abx, "Cohoused", Mouse), paste(Abx, 'Solitary', Mouse)))
  
p <- j %>%
  ggplot(aes(x = Hour, y = shannon_val, color = Abx)) +
    geom_point() +
    stat_smooth() +
    scale_color_manual(values = c('red','black')) +
    geom_rect(aes(xmin = 0, xmax = 120, ymin = -Inf, ymax = Inf), alpha = 0.3, fill = 'red', color = NA, data = j %>% group_by(Mouse) %>% filter(row_number() == 1 & Abx == 'Abx')) + 
    facet_wrap(~Mouse, ncol = 3) +
    labs(x = 'Hour', y = 'Shannon diversity') +
    th

plot_it(p, name = 'figShannonOverTime', dir = image_dir, width = 6, height = 8)

p

```

(ref:figShannonOverTime) Figure `r refs$figShannonOverTime`

`r pagebreak` ![](images\\figShannonOverTime.png)

**(ref:figShannonOverTime):** Shannon diversity over time within mice.

```{r, include = FALSE}

p <- j %>%
  ggplot(aes(x = Hour, y = rich, color = Abx)) +
    geom_point() +
    stat_smooth() +
    scale_color_manual(values = c('red','black')) +
    geom_rect(aes(xmin = 0, xmax = 120, ymin = -Inf, ymax = Inf), alpha = 0.3, fill = 'red', color = NA, data = j %>% group_by(Mouse) %>% filter(row_number() == 1 & Abx == 'Abx')) + 
    facet_wrap(~Mouse, ncol = 3) +
    labs(x = 'Hour', y = 'OTU richness') +
    th

plot_it(p, name = 'figRichOverTime', dir = image_dir, width = 6, height = 8)

p

```

(ref:figRichOverTime) Figure `r refs$figRichOverTime`

`r pagebreak` ![](images\\figRichOverTime.png)

**(ref:figRichOverTime):** OTU richness over time within mice.

```{r, include = FALSE}

p <- j %>%
  ggplot(aes(x = Hour, y = evenness, color = Abx)) +
    geom_point() +
    stat_smooth() +
    scale_color_manual(values = c('red','black')) +
    geom_rect(aes(xmin = 0, xmax = 120, ymin = -Inf, ymax = Inf), alpha = 0.3, fill = 'red', color = NA, data = j %>% group_by(Mouse) %>% filter(row_number() == 1 & Abx == 'Abx')) + 
    facet_wrap(~Mouse, ncol = 3) +
    labs(x = 'Hour', y = 'Community evenness') +
    th

plot_it(p, name = 'figEvennessOverTime', dir = image_dir, width = 6, height = 8)

p

```

(ref:figEvennessOverTime) Figure `r refs$figEvennessOverTime`

`r pagebreak` ![](images\\figEvennessOverTime.png)

**(ref:figEvennessOverTime):** OTU community evenness (i.e., equitability, i.e., Shannon diversity H / log(OTU richness))


```{r, include = FALSE}

#I am assuming that OTUs that are absent from a given mouse in the last three samples collected during antibiotic dosing (i.e., at hours 63, 73 and 97) are not in the mouse. This window can be tweaked.

#set the window. The edges are included.
window0 <- 63 
window1 <- 120

j <- x %>%
  group_by(Mouse, Abx, Pair, otu) %>%
  summarise(
    abunPre = if (unique(Abx) == 'Y') sum(abun[Hour <= window0]) else NA,
    abun0 = if (unique(Abx) == 'Y') sum(abun[Hour >= window0 & Hour <= window1]) else sum(abun[Hour < window1]),
    abun1 = sum(abun[Hour > window1]),
    statusAB = if (!is.na(abunPre) & abunPre == 0 & abun0 == 0) 'absent' else 'persistent',
    statusAB = if (!is.na(abunPre) & abunPre > 0 & abun0 == 0) 'extinct' else statusAB,
    ETD = if (statusAB == 'extinct') max(Hour[Hour <= window0 & abun > 0]) else NA,
    ETD = if (statusAB == 'persistent') Inf else ETD,
    status0 = if (abun0 > 0) 'present' else 'absent',
    status1 = if (sum(abun[Hour > window1] > 0)) 'present' else 'absent',
    status = if (status0 == 'absent' & status1 == 'absent') 'absent' else 'persistent',
    status = if (status0 == 'absent' & status1 == 'present') 'immigrant' else status,
    status = if (status0 == 'present' & status1 == 'absent') 'extinct' else status,
    ETA = if (status == 'immigrant') min(Hour[Hour > 97 & abun > 0]) else NA) %>%
  group_by(Pair, otu) %>%
  mutate(
    abun_pair = ifelse(!is.na(Pair) & Abx == 'Y', abun0[Abx == 'N'], NA),
    abun_pair = ifelse(!is.na(Pair) & Abx == 'N', abun0[Abx == 'Y'], abun_pair))
    
j <- x %>%
  group_by(Mouse, Abx, Pair, otu) %>%
  summarise(
    abun0 = sum(abun[Hour < window0]),
    abun1 = sum(abun[Hour >= window0 & Hour <= window1]),
    abun01 = abun0 + abun1,
    abun2 = sum(abun[Hour > window1]),
    status_treatment = if (abun01 == 0) 'absent' 
                         else if (abun0 > 0 & abun1 == 0) 'extinct' 
                           else 'persistent',
    ETD = if (status_treatment == 'extinct') max(Hour[Hour <= window0 & abun > 0]) 
            else if (status_treatment == 'persistent') Inf 
              else NA,
    status_recovery = if (abun01 == 0 & abun2 == 0) 'absent' 
               else if (abun1 == 0 & abun2 > 0) 'immigrant'
                 else if (abun01 > 0 & abun2 == 0) 'extinct'
                   else 'persistent',
    immigrant_type = if (status_recovery != 'immigrant') 'Not an immigrant'
                       else if (abun01 > 0) 'recovered'
                         else 'new colonist',
    ETA = if (status_recovery == 'immigrant') min(Hour[Hour > 97 & abun > 0]) 
            else NA) %>%
  group_by(Pair, otu) %>%
  mutate(
    abun_pair = ifelse(!is.na(Pair) & Abx == 'Y', abun01[Abx == 'N'], NA),
    abun_pair = ifelse(!is.na(Pair) & Abx == 'N', abun01[Abx == 'Y'], abun_pair))

p <- j %>% 
  filter(status_recovery == 'immigrant' & Abx == 'Y' & abun_pair > 0) %>%
  mutate(time_to_arrival = ETA - 120) %>%
  ggplot(aes(x = abun_pair, y = time_to_arrival)) + 
    geom_jitter(alpha = 0.25, color = 'red', height = 5, width = 0) + 
    scale_x_log10() + 
    stat_smooth(color = 'black', method = 'lm') +
    stat_smooth(color = 'black', lty = 3) +
    labs(x = "OTU relative abundance in cohoused control mouse", y = 'Hours before OTU arrival in Abx treated mouse') +
    th

plot_it(p, name = 'figPropPressure', dir = image_dir, width = 6.5, height = 5)

p 

```

(ref:figPropPressure) Figure `r refs$figPropPressure`

`r pagebreak` ![](images\\figPropPressure.png)

**(ref:figPropPressure):** A test of the potential role of mass effects. The hypothesis is that, all else being equal, the abundance of a taxon in the regional pool (i.e., the cohoused mouse) is going to correlate positively with time to arrival (i.e., successful colonization) in the treated mouse. This hypothesis is supported. However, we can't be sure if the hypothesis is true because the taxa that are more abundant taxa in the control mouse may simply be better suited for the gut environment, and thus are faster to colonize the antibiotic-treated mouse. So, the analysis has it's limitations, but I still think it's worth noting.

```{r, include = FALSE}

p <- x %>%
  ungroup() %>%
  left_join(j[, c('Mouse','otu','status_recovery','immigrant_type')], by = c('Mouse','otu')) %>%
  filter(!is.na(Pair) & abun > 0 & status_recovery != 'absent') %>%
  arrange(status_recovery) %>%
  mutate(status_recovery = ifelse(status_recovery == 'immigrant', immigrant_type, status_recovery)) %>%
  mutate(OTU_status = c('Extinct by end of Abx treatment (hour 120 at the latest); does not return during recovery period', 
                        'New colonist, arrived during recovery period (after hour 120)',
                        'Present throughout experiment',
                        'Extinct during end of Abx treatment (between Hours 63-120); returns during recovery period (after Hour 144)')[match(as.numeric(factor(status_recovery, levels = sort(unique(status_recovery)))), c(1:4))],
         Pair = paste('Pair', Pair),
         Abx = ifelse(Abx == 'Y', 'Abx', 'No abx'))

p <- ggplot(p, aes(x = as.factor(Hour), y = abun, fill = OTU_status)) +
    geom_bar(stat='identity', alpha = 0.8) +
    geom_rect(xmin = 0, xmax = 14, ymin = -Inf, ymax = 0, fill = 'red', data = filter(p, Abx == 'Abx')) +
    facet_grid(Pair~Abx) +
    scale_fill_manual(name = '', values = c('#a6611a','#dfc27d','#80cdc1','#018571')) +
    #scale_fill_discrete(name = '') +
    labs(x = 'Hour: {0,4,8,12,16,20,23,25,39,49,63,73,97,120,144,168,192,216,240,312}', y = 'Rarefied abundance') +
    guides(fill=guide_legend(nrow=4,byrow=TRUE, position = 'bottom')) +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      axis.text.x=element_blank(),
      legend.position = 'bottom')

plot_it(p, name = 'figOTUAbunByAbxStatus', dir = image_dir, width = 7, height = 8)

p 

```

(ref:figOTUAbunByAbxStatus) Figure `r refs$figOTUAbunByAbxStatus`

`r pagebreak` ![](images\\figOTUAbunByAbxStatus.png)

**(ref:figOTUAbunByAbxStatus):** Rarefied OTU abundances, colored by OTU status. This figure is to compare mice treated by antibiotics to untreated mice. 

```{r, include = FALSE}

p <- x %>%
  ungroup() %>%
  left_join(j[, c('Mouse','otu','status_recovery','immigrant_type')], by = c('Mouse','otu')) %>%
  filter(Abx == 'Y' & abun > 0 & status_recovery != 'absent' & Mouse != 1) %>%
  arrange(status_recovery) %>%
  mutate(status_recovery = ifelse(status_recovery == 'immigrant', immigrant_type, status_recovery)) %>%
  mutate(OTU_status = c('Extinct by end of Abx treatment (hour 120 at the latest); does not return during recovery period', 
                        'New colonist, arrived during recovery period (after hour 120)',
                        'Present throughout experiment',
                        'Extinct during end of Abx treatment (between Hours 63-120); returns during recovery period (after Hour 144)')[match(as.numeric(factor(status_recovery, levels = sort(unique(status_recovery)))), c(1:4))],
         Mouse = paste(ifelse(Cohoused == 'Y', 'Cohoused', 'Not cohoused'), "- Mouse", Mouse),
         Abx = ifelse(Abx == 'Y', 'Abx', 'No abx')) %>%
  ggplot(aes(x = as.factor(Hour), y = abun, fill = OTU_status)) +
    geom_bar(stat='identity', alpha = 0.8) +
    geom_rect(xmin = 0, xmax = 14, ymin = -Inf, ymax = 0, fill = 'red') +
    facet_wrap(~Mouse, ncol = 2) +
    scale_fill_manual(name = '', values = c('#a6611a','#dfc27d','#80cdc1','#018571')) +
    #scale_fill_discrete(name = '') +
    labs(x = 'Hour: {0,4,8,12,16,20,23,25,39,49,63,73,97,120,144,168,192,216,240,312}', y = 'Rarefied abundance') +
    guides(fill=guide_legend(nrow=4,byrow=TRUE, position = 'bottom')) +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      axis.text.x=element_blank(),
      legend.position = 'bottom')

plot_it(p, name = 'figOTUAbunByCageStatus', dir = image_dir, width = 7, height = 8)

p 

```

(ref:figOTUAbunByCageStatus) Figure `r refs$figOTUAbunByCageStatus`

`r pagebreak` ![](images\\figOTUAbunByCageStatus.png)

**(ref:figOTUAbunByCageStatus):** Rarefied OTU abundances, colored by OTU status. This figure is to compare mice in cohousing environments during the recovery process, with mice caged alone during the recovery process.

```{r, include = FALSE}

p <- x %>%
  ungroup() %>%
  left_join(j[, c('Mouse','otu','status_recovery','immigrant_type')], by = c('Mouse','otu')) %>%
  filter(!is.na(Pair) & abun > 0 & status_recovery != 'absent') %>%
  arrange(status_recovery) %>%
  mutate(status_recovery = ifelse(status_recovery == 'immigrant', immigrant_type, status_recovery)) %>%
  mutate(OTU_status = c('Extinct by end of Abx treatment (hour 120 at the latest); does not return during recovery period', 
                        'New colonist, arrived during recovery period (after hour 120)',
                        'Present throughout experiment',
                        'Extinct during end of Abx treatment (between Hours 63-120); returns during recovery period (after Hour 144)')[match(as.numeric(factor(status_recovery, levels = sort(unique(status_recovery)))), c(1:4))],
         Pair = paste('Pair', Pair),
         Abx = ifelse(Abx == 'Y', 'Abx', 'No abx'))

p <- ggplot(p, aes(x = as.factor(Hour), fill = OTU_status)) +
    geom_bar(alpha = 0.8) +
    geom_rect(xmin = 0, xmax = 14, ymin = -Inf, ymax = 0, fill = 'red', alpha = 0.2, data = filter(p, Abx == 'Abx')) +
    facet_grid(Pair~Abx) +
    scale_fill_manual(name = '', values = c('#a6611a','#dfc27d','#80cdc1','#018571')) +
    #scale_fill_discrete(name = '') +
    labs(x = 'Hour: {0,4,8,12,16,20,23,25,39,49,63,73,97,120,144,168,192,216,240,312}', y = 'OTU richness') +
    guides(fill=guide_legend(nrow=4,byrow=TRUE, position = 'bottom')) +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      axis.text.x=element_blank(),
      legend.position = 'bottom')

plot_it(p, name = 'figOTURichByAbxStatus', dir = image_dir, width = 7, height = 8)

p 

```

(ref:figOTURichByAbxStatus) Figure `r refs$figOTURichByAbxStatus`

`r pagebreak` ![](images\\figOTURichByAbxStatus.png)

**(ref:figOTURichByAbxStatus):** OTU richness from communities rarefied to 5000 seqeunces, colored by OTU status. This figure is to compare mice treated by antibiotics to untreated mice. 

```{r, include = FALSE}


p <- x %>%
  ungroup() %>%
  left_join(j[, c('Mouse','otu','status_recovery','immigrant_type')], by = c('Mouse','otu')) %>%
  filter(Abx == 'Y' & abun > 0 & status_recovery != 'absent' & Mouse != 1) %>%
  arrange(status_recovery) %>%
  mutate(status_recovery = ifelse(status_recovery == 'immigrant', immigrant_type, status_recovery)) %>%
  mutate(OTU_status = c('Extinct by end of Abx treatment (hour 120 at the latest); does not return during recovery period', 
                        'New colonist, arrived during recovery period (after hour 120)',
                        'Present throughout experiment',
                        'Extinct during end of Abx treatment (between Hours 63-120); returns during recovery period (after Hour 144)')[match(as.numeric(factor(status_recovery, levels = sort(unique(status_recovery)))), c(1:4))],
         Mouse = paste(ifelse(Cohoused == 'Y', 'Cohoused', 'Not cohoused'), "- Mouse", Mouse),
         Abx = ifelse(Abx == 'Y', 'Abx', 'No abx')) %>%
  ggplot(aes(x = as.factor(Hour), fill = OTU_status)) +
    geom_bar(alpha = 0.8) +
    geom_rect(xmin = 0, xmax = 14, ymin = -Inf, ymax = 0, fill = 'red') +
    facet_wrap(~Mouse, ncol = 2) +
    scale_fill_manual(name = '', values = c('#a6611a','#dfc27d','#80cdc1','#018571')) +
    #scale_fill_discrete(name = '') +
    labs(x = 'Hour: {0,4,8,12,16,20,23,25,39,49,63,73,97,120,144,168,192,216,240,312}', y = 'OTU richness') +
    guides(fill=guide_legend(nrow=4,byrow=TRUE, position = 'bottom')) +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      axis.text.x=element_blank(),
      legend.position = 'bottom')


plot_it(p, name = 'figOTURichByCageStatus', dir = image_dir, width = 7, height = 8)

p 

```

(ref:figOTURichByCageStatus) Figure `r refs$figOTURichByCageStatus`

`r pagebreak` ![](images\\figOTURichByCageStatus.png)

**(ref:figOTURichByCageStatus):** OTU richness from communities rarefied to 5000 seqeunces, colored by OTU status. This figure is to compare mice in cohousing environments during the recovery process, with mice caged alone during the recovery process.

```{r, include = FALSE}

st <- x %>%
  group_by(otu, Day = floor(Hour/24)) %>%
  filter(sum(abun[Abx == 'Y'] > 0) & sum(abun[Abx == 'N'] > 0)) %>%
  do(mod = t.test(abun~Abx, data = .)) %>%
  mutate(
    pval = mod$p.value,
    df = mod$parameter,
    tval = mod$statistic,
    mean_No_Abx = mod$estimate[[1]],
    mean_Yes_Abx = mod$estimate[[2]],
    diff = mean_No_Abx - mean_Yes_Abx)

j <- ungroup(st) %>% 
  group_by(otu) %>% 
  summarise(sigs = sum(pval < 0.05), sign = sum(diff)) %>% 
  arrange(desc(abs(sign))) %>%
  do(data.frame(winners = 
    head(.$otu[.$sign < 0]),
    losers = head(.$otu[.$sign > 0]))) %>%
  gather(status, otu)

p <- x %>%
  left_join(tax) %>%
  filter(otu %in% j$otu) %>%
  ungroup() %>%
  mutate(
    label = paste(Family, 'OTU', as.numeric(as.factor(otu))),
    Abx = ifelse(Abx == 'N', 'No abx', 'Abx')) %>%
  left_join(j) %>%
  ggplot(aes(x = Hour, y = abun + 1, color = Abx)) + 
  annotate(geom = 'rect', xmin = 1, xmax = 120, ymin = 0, ymax = Inf, alpha = 0.3, fill = 'red') + 
  geom_point() + 
  stat_smooth() +
  #geom_text(label = "Antibiotics", x = 60, aes(y = max(abun))) +
  #geom_text(label = "Recovery", x = 216, aes(y = max(abun))) +
  scale_y_log10() +
  scale_color_manual(values = c('red','black'), name = '') +
  facet_wrap(~label) + 
  theme_bw() +
  theme(
    panel.grid.minor = element_blank(),
    panel.grid.major = element_blank(),
    panel.background = element_blank(),
    legend.position = 'bottom')

plot_it(p, name = 'figMostAffectedOTUs', dir = image_dir, width = 9, height = 8)

p
```

(ref:figMostAffectedOTUs) Figure `r refs$figMostAffectedOTUs`

`r pagebreak` ![](images\\figMostAffectedOTUs.png)

**(ref:figMostAffectedOTUs):** Trends in abundances of 12 taxa that differ significantly between antibiotic treated mice and control mice. The taxa were selected because they exhibited the greatest differences in abundances between treatments. The translucent red square highlights the window of antibiotic treatment.

```{r, include = FALSE}

iwant <- c('Copies_16S','Forms_biofilms','Gene_number','Obligate_anaerobe','Sporulation')

tmp <- x %>%
  left_join(traits_wide[ ,c('otu', iwant)], by = 'otu') %>%
  gather(trait, val, .dots = iwant) %>%
  group_by(Mouse, Abx, Cohoused, Hour, trait) %>%
  filter(Cohoused != 'dead') %>%
  summarise(CWM = weighted.mean(val, w = abun)) %>%
  ungroup() %>%
  mutate(Abx = ifelse(Abx == 'Y', 'Abx', 'No abx'),
         Cohoused = ifelse(Cohoused == 'Y', 'Cohoused', 'Solitary'))

p <- ggplot(tmp, aes(x = Hour, y = CWM, color = Abx)) +
    geom_rect(aes(xmin = 0, xmax = 120, ymin = -Inf, ymax = Inf), alpha = 0.3, fill = 'red', color = NA, data =  group_by(tmp, trait, Abx) %>% filter(row_number() == 1 & Abx == 'Abx')) +
    stat_smooth(aes(group = Abx), se = FALSE, size = 1.5) +
    stat_smooth(aes(lty = Cohoused, group = Mouse), geom="line", se = FALSE, alpha = 0.6) +
    scale_color_manual(values = c('red','black'), name = '') +
    scale_linetype_discrete(name = '') +
    facet_grid(trait~Abx, scales = 'free') +
    guides(lty=guide_legend(override.aes = list(color = 'black', alpha = 1))) +
    th

plot_it(p, name = 'figFiveTraitPatterns', dir = image_dir, width = 6, height = 7.5)

p

```

(ref:figFiveTraitPatterns) Figure `r refs$figFiveTraitPatterns`

`r pagebreak` ![](images\\figFiveTraitPatterns.png)

**(ref:figFiveTraitPatterns):** Shifts in community weighted means of mouse gut microbial communities during and after antibiotic treatment (left) or in untreated control mice (right). Thick colored lines show average CWM trait values across mice, whereas thin colored lines show CWM trends for individual mice.

```{r, include = FALSE}

ggtax <- readRDS(paste0(wd_dat, 'ggtax.RDS')) %>%
  mutate(otu = as.integer(gsub('otu', '', otu)))

tmp <- traits_wide %>%
  gather(trait, val, -otu) %>%
  filter(!is.na(val)) %>%
  left_join(ggtax, by = 'otu') %>%
  filter(Species != 'unclassified')

tmp_abuns <- x %>% ungroup() %>% group_by(otu) %>% summarise(abun = sum(abun))  

#I drop traits that are already dependent on inferred trait data, from Bugbase
j <- traits_wide %>%
  select(-Forms_biofilms, -Obligate_anaerobe, -Gram_positive) %>%
  gather(trait, val, -otu) %>%
  distinct(otu, trait) %>%
  inner_join(tax, by = 'otu') %>%
  mutate(
    Species = paste(trait, Family, Genus, Species) %in% paste(tmp$trait, tmp$Family, tmp$Genus, tmp$Species),
    Genus = paste(trait, Family, Genus) %in% paste(tmp$trait, tmp$Family, tmp$Genus),
    Family = paste(trait, Family) %in% paste(tmp$trait, tmp$Family),
    Order = paste(trait, Order) %in% paste(tmp$trait, tmp$Order),
    Class = paste(trait, Class) %in% paste(tmp$trait, tmp$Class)
  ) %>%
  rename(C = Class, O = Order, F = Family, G = Genus, S = Species) %>%
  gather(level, coverage, C, O, F, G, S) %>%
  inner_join(tmp_abuns, by = 'otu') %>%
  mutate(level = factor(level, levels = rev(c('C','O','F','G','S')))) %>%
  group_by(trait, level)

# individuals with trait data by trait and level
#j %>% 
#  group_by(trait, level) %>% 
#  summarise(abun = sum(abun[coverage])) %>%
#  spread(level, abun)

#percent individuals with trait data by trait and level
#j %>% 
#  group_by(trait, level) %>% 
#  summarise(abun = round(sum(abun[coverage]) / sum(abun), 2)) %>%
#  spread(level, abun)

# summary across traits
#j %>% 
#  group_by(level) %>% 
#  summarise(abun = sum(abun[coverage])) %>%
#  mutate(abun_percent = abun / max(abun))

# numbers of OTUs with trait data by trait and level
#j %>% 
#  group_by(trait, level) %>% 
#  summarise(rich = sum(coverage)) %>%
#  spread(level, rich)

# numbers of OTUs with trait data by level
#j %>% 
#  group_by(trait, level) %>% 
#  summarise(rich = round(sum(coverage) / length(coverage), 2)) %>%
#  spread(level, rich)

# summary across traits
#j %>% 
#  group_by(level) %>% 
#  summarise(rich = sum(coverage)) %>%
#  mutate(rich_percent = rich / max(rich))

p <- j %>%
  summarise(coverage = sum(abun[coverage]) / sum(abun)) %>%
  filter(trait %in% iwant) %>%
  ungroup() %>%
  mutate(trait = ifelse(trait %in% names(trait_names), trait_names[match(trait, names(trait_names))], trait)) %>%
  ggplot(aes(x = level, y = coverage)) +
    geom_bar(stat = 'identity') +
    facet_wrap(~trait) +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      panel.background = element_blank(),
      legend.position = 'none') +
    labs(x = 'Maximum level of phylogenetic distance allowed when inferring trait data', y = 'Proportion of community with available trait data')

plot_it(p, name = 'figCoverage', dir = image_dir, width = 6, height = 3.5)

p

```

(ref:figCoverage) Figure `r refs$figCoverage`

`r pagebreak` ![](images\\figCoverage.png)

**(ref:figCoverage):** 
Unknown trait data were inferred by averaging trait values over descending subtrees in the greengenes phylogeny. That is, for each unknown trait value, the predicted value was set to the arithmetic average state of all tips descending from that node. On the x-axis, S = Species, G = Genus, F = Family, O = Order, C = Class. Missing trait data for three traits (the functional potential to form biofilms, obgligate anaerobe status, and Gram-positive status) were inferred for all OTUs in the greengenes phylogeny in Ward et al. 2017, and are therefore not reported in this figure.


`r pagebreak` Works Cited